### ISSUE TITLE
[Feature Request]: Playlists/m3u files

### ISSUE THREAD

**Issue by TomTurbine** [PCSX2/pcsx2#6696]
### Describe the Bug

When you try to load up an M3U file it gives the error "Failed to Initialize CDVD"

If I launch the game manually it works just fine but no m3u file will load.

### Reproduction Steps

Just attempt to load up any m3u file.

### PCSX2 Revision

Build 1.7.3115   (AVX2 - Qt)

### Operating System

Windows 10 (64bit)

**Comment by stenzek** [comment]
Not supported. This is effectively a feature request.

**Comment by TomTurbine** [comment]
If it's not supported than this goes from a bug report to an error report.

PCSX2 shows support for M3U files when you click "Start File" and is listed as a "Playlist" and when you filter by type it gives you the option of filtering for playlists. But doesn't show up there when you do.

PCSX2 shows all the support options for loading the m3u file and even gives the option to launch one manually.

**Comment by TomTurbine** [comment]
[https://ibb.co/c6Hp2jH](url)

**Comment by TomTurbine** [comment]
Might be a planned feature that isn't yet implemented. If so, might want to disable them options till it is to prevent confusion.

**Comment by stenzek** [comment]
PRs welcome, but it seems silly to remove it from the UI when it might be implemented at a future date.

**Comment by TomTurbine** [comment]
Was mainly thinking of the confusion it will cause people like myself who are trying to figure out why it would give the options for an m3u and then not use it. Makes them wonder if they were screwing up or a bug.

Just disabling it in the UI till it is ready removes the idea that it is implemented yet. And removing it from the UI is trivial.

Or just put a comment beside those options saying "Not yet implemented" which would 100% remove all doubt from them till then.

**Comment by jeltaqq** [comment]
I was just curious. How do you use M3U files in a PS scene?

**Comment by TomTurbine** [comment]
Well, there are some multi-disc games and instead of loading up the discs directly just load up the playlist and enter a button to swap discs between them when the game asks for it.

**Comment by TomTurbine** [comment]
It really depends on how it is setup. If you have it setup where you are using a front end like Emulationstation or Launchbox, having a hotkey setup to swap disc and having it load from the m3u file instead of straight from the disc is a huge convenience.

Have a PS5 controller setup and trying to have the Microphone button hotkeyed to Swap discs for when I play games like Star Ocean.

Also, a portable hard drive I am trying to setup for the kids and don't know how to setup a relative directory to the games folders in the emulator itself.

**Comment by jeltaqq** [comment]
I'm looking it working on duckstation and it works as a folder (like Groups on Nintendo Switch) sounds like fun.

**Comment by Deozaan** [comment]
I just found my way here because I was confused about why PCSX2 listed itself as supporting `Playlists (*.m3u)` when clicking `System -> Start File` but it not working and my `.m3u` files not showing up in the game list when scanning my game directory.

**Comment by stenzek** [comment]
Yeah, that should be removed. I'm not planning on implementing it any time soon, since it's arguably even less useful now that there's the big picture UI.

**Comment by fgmenth** [comment]
don't forget to remove it from the Filter too

**Comment by RedDevilus** [comment]
It doesn't show up in the gamelist anyway and PR removed the Open File extension listing.

**Comment by chez2212** [comment]
gotta LOVE when developers are like, "this feature isn't useful because I said so" and don't listen to individuals saying YES IT IS, probably would even be that hard to implement, but because the developer thinks it is useless because they think their way is the only way that matters, they ignore people asking for it.

I do not use ps2 front end, i use emulation station on a system that has no keyboard or mouse to easily get into the menus to change a disc, a feature just like the one in ps1, where you can load a m3u, then use a hotkey to change the disc when the game asks for it, would make ease better for those games that use multiple discs, like the following list, and i cant even say if this list is full or not, these are just the ones I have.

Armored Core: Nexus
Cy Girls
Devil May Cry 2
Grandia III
Metal Gear Solid 3: Subsistence
Onimusha: Dawn of Dreams
Sakura Wars: So Long, My Love
Shadow Hearts: Covenant
Star Ocean: Till the End of Time
Xenosaga Episode II
Xenosaga Episode III

I can understand if it was 1 game that had multiple discs and saying it wasn't worth the effort, but when there are multiple games that have more then 1 disc, and multiple people asking for the feature, for a feature I'm sure it wouldn't be that hard to add, seems arrogant to say its a feature that isn't needed, honestly I don't know C++, I probably could figure it out after some time, so I may look into it myself, as I do know many other languages from scripting to some programing, I'm sure it would be a feature that wouldn't take a huge amount of time to implement  for someone who knows the source code and C++, I think you should just add it even if you think its not useless.

**Comment by JordanTheToaster** [comment]
Reopening as it is a valid feature request.

**Comment by WestonHanners** [comment]
Just came across this, if there is no m3u support, is there some alternate method of bundling multi-disc games up so that front ends like Emulation Station can see them as one game but still let PCSX2 see both disks? 

I know PS1 emulators support multi-disc CHD's maybe something like that?

**Comment by chez2212** [comment]
> Just came across this, if there is no m3u support, is there some alternate method of bundling multi-disc games up so that front ends like Emulation Station can see them as one game but still let PCSX2 see both disks?
> 
> I know PS1 emulators support multi-disc CHD's maybe something like that?

I don't think there there is a method for a multi-disc CHD at all, there is methods for multi-disk bin and cue as far as I know of, but never messed with it, for PS1 there is pbp format as well, but outside ppsspp, i dont think any ps1 emulator supports it, as its a psp format and only works for ps1 games, and I wouldnt want to merge multiple discs into one file anyways, I personally dislike huge files like that. Anyways, most emulators with systems that have multiple discs can read and use m3u files and have hotkeys to switch the disc when needed, only other one that I know of that doesn't support m3u files is ppsspp, but there is only 1 psp multi disc game that I am aware of, and their system detects it somehow and adds a menu item to select the disc as soon as the game asks for the 2nd disc, unfortunately that still runs into the same issue where a system with the setup without a keyboard and mouse can't easily just switch the disc, would it be nice if they also added m3u support, but i understand, 1 game out of how many psp games, granted there may be more, but not many. I found after my comment earlier that there is actually a lot more multi disc PS2 games then I was aware of, a lot more then I even listed, quite a few actually, so I think the ability to read m3u files with a hotkey to switch the disc would be a very useful addition to PCSX2, especially for those who use a setup with just a controller like I do, its not to say its hard to hook up a keyboard and mouse, but its a huge inconvenience, and when using a front end, it makes things a lot cleaner, I mean, right now I have multiple entries in Emulation Station DE for each disc of the same games, not only does this look pretty ugly imho, it also ends up taking up more room, while not a lot, it still takes up room, since each entry has its own copy of the images and preview video, where when I use m3u in psx, gamecube, dreamcast, saturn, and other disc systems, it allows me to keep it clean, as the file structure is as simple as a m3u file and all the discs in a folder with the same name as the m3u file, for example

Final Fantasy VII.m3u\Final Fantasy VII (Disc 1).chd
Final Fantasy VII.m3u\Final Fantasy VII (Disc 2).chd
Final Fantasy VII.m3u\Final Fantasy VII (Disc 3).chd
Final Fantasy VII.m3u\Final Fantasy VII.m3u

This keeps things clean and organized, and only counts as one entry in ES-DE. and it isn't just ES-DE this would work with, any front end would have the same issue and benefit from the addition to m3u file support, and there are quite a few front ends individuals use.

**Comment by TomTurbine** [comment]
> Just came across this, if there is no m3u support, is there some alternate method of bundling multi-disc games up so that front ends like Emulation Station can see them as one game but still let PCSX2 see both disks?
> 
> I know PS1 emulators support multi-disc CHD's maybe something like that?

Currently not on PSX2. While you can make an M3U file for Duckstation to load and then have the front end of your choice hide showing the actual ISO/CHDs. For PCSX2, without any multidisc support of any kind, you have to have each disc to show up on your front end to load.

Can understand why M3U support is such a low priority since the number of multi-disc games are so low. But this is also something that would be trivial to implement compared to all the actual emulation work they do. But also, since there isn't a single PS2 game I know of that has more than 2 discs, the whole disc swap thing would be that much easier I would think.

Could potentially even code those games to auto-swap discs when the games ask for it potentially since you know there is exactly only 1 other disc it will be waiting for in that M3U file.

**Comment by TomTurbine** [comment]
> Just came across this, if there is no m3u support, is there some alternate method of bundling multi-disc games up so that front ends like Emulation Station can see them as one game but still let PCSX2 see both disks?
> 
> I know PS1 emulators support multi-disc CHD's maybe something like that?

Currently not on PSX2. While you can make an M3U file for Duckstation to load and then have the front end of your choice hide showing the actual ISO/CHDs. For PCSX2, without any multidisc support of any kind, you have to have each disc to show up on your front end to load.

Can understand why M3U support is such a low priority since the number of multi-disc games are so low. But this is also something that would be trivial to implement compared to all the actual emulation work they do. But also, since there isn't a single PS2 game I know of that has more than 2 discs, the whole disc swap thing would be that much easier I would think.

Could potentially even code those games to auto-swap discs when the games ask for it potentially since you know there is exactly only 1 other disc it will be waiting for in that M3U file.

**Comment by Cainam-git** [comment]
I also need this feature specifically for ESDE usage. The workaround right now is to hide the second disc and deliberately change the metadata of disc 1 to be the game name without disc 1/disc 2. This isn't a clean method because some games do not even ask you to swap out the disc (e.g. Devil May Cry 2 where you play as a different protagonist depending on the disc you insert, both discs more or less contain the 'full' game). You would completely miss the existence of the second disc this way.

**Comment by quidamphx** [comment]
Any thoughts on this feature being implemented? It's a supported feature on most other disc emulators, and while I know that doesn't mean PCSX2 needs to follow suit, it's something I want and actively use when possible. Much cleaner to have single game entries, just like seeing it on a shelf. No one has separate game cases for disc 1 and disc 2, etc.

**Comment by Neo-Retro** [comment]
any news for m3u? i don't like the look and feel with double cover in showcase. duckstation do this great and work fine, a similar function for PCSX2 would be awesome. thank you in advance!

**Comment by legluondunet** [comment]
it's a very useful feature when you use Retroarch or Emulationstation-DE, this avoid to have several time the same game shown in your list.

**Comment by ShizCalev** [comment]
> But also, since there isn't a single PS2 game I know of that has more than 2 discs, the whole disc swap thing would be that much easier I would think.

Metal Gear Solid 3: Subsistence has 3 discs, and The Fear has 4.

**Comment by nfriedly** [comment]
I put up a PR last night with a very basic implementation of this - it parses .m3u files and then always loads the first disc in the list. 

I did some light testing of the first windows build with MGS3: Subsistence (.chd files and a .m3u that I made myself), but I'd appreciate help in testing it further:

* What happens when automatically loading a save state that is expecting the second disc?(I don't have any saves that are beyond the first disk of any multidisc games.)
* Does it work correctly with batocera and other frontends?
* Can you break it with a bad .m3u file beyond the expected error messages?

The PR is at https://github.com/PCSX2/pcsx2/pull/14271 and my fork has builds of it that you can download at https://github.com/nfriedly/pcsx2/actions - select the latest windows/linux/mac build, the scroll down to the artifacts section and download one of the builds to test. (Note that you have to be logged into github to see them, scrolling doesn't work when your mouse is over the light-grey section, and the builds will be automatically deleted eventually.)

I think some additional work is warranted - I was thinking it would be good to add every file in the list to the System-> Change Disc menu, although that might be in a future PR. I'm a little more hazy on other improvements, although I put some ideas in the PR.

**Comment by proletariado** [comment]
Sorry for bumping this but is there any reason why #14271 hasn't been merged yet?

**Comment by nfriedly** [comment]
@proletariado I assume it's because the PCSX2 team is flooded with PRs and no one has had the time to review it yet. 

Also, my PR was mostly AI-written, and as someone who reviews a lot of code in my day job, I'm willing to admit that the AI-generated ones can be harder to review. 

Finally, while the PR covers the bare basics of enabling .m3u support, it doesn't interact with save states very well. I have some local changes to improve that, but they're not finished yet. I wouldn't blame a reviewer for wanting to wait for me to finish that before starting their review.
